### PR TITLE
New Crypto Library written in Rust

### PR DESCRIPTION
**Pull request by sallar**

### CODE REVIEW

File: android/app/src/main/java/com/buttercup/Crypto.java [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+import com.facebook.react.bridge.ReactMethod;
+import com.facebook.react.bridge.Promise;
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.facebook.react.bridge.ReactContextBaseJavaModule;
+
+import java.net.URLDecoder;
+import java.lang.String;
+
+public class Crypto extends ReactContextBaseJavaModule {
+
```
Comment (perry-mitchell) [inline android/app/src/main/java/com/buttercup/Crypto.java]: Can we remove the empty line?

File: android/app/src/main/java/com/buttercup/Crypto.java [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+            return;
+        }
+        promise.reject("Generating IV failed.");
+    }
+
+    @Override
+    public String getName() {
+        return "Crypto";
+    }
+
```
Comment (perry-mitchell) [inline android/app/src/main/java/com/buttercup/Crypto.java]: And here too?

File: ios/Crypto.m [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+// The React Native bridge needs to know our module
+RCT_EXPORT_MODULE()
+
++ (BOOL)requiresMainQueueSetup
+{
+    return YES;
+}
+
+RCT_EXPORT_METHOD(pbkdf2:(NSString *)password:(NSString *)salt:(int)iterations:(int)bits:(RCTPromiseResolveBlock)resolve:(RCTPromiseRejectBlock)reject) {
+
```
Comment (perry-mitchell) [inline ios/Crypto.m]: Blank line..

File: ios/Crypto.m [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+        reject(
+            @"pbkdf2_failed",
+            @"Key Derivation failed.",
+            [BCHelpers newErrorObject]
+        );
+    }
+}
+
+RCT_EXPORT_METHOD(encryptText:(NSString *)data:(NSString *)key:(NSString *)salt:(NSString *)ivHex:(NSString *)hmacHexKey:(RCTPromiseResolveBlock)resolve:(RCTPromiseRejectBlock)reject) {
+
```
Comment (perry-mitchell) [inline ios/Crypto.m]: And here..

File: ios/Crypto.m [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+#include <sys/sysctl.h>
+#import "BCHelpers.h"
+
+@implementation Crypto
+
+// The React Native bridge needs to know our module
+RCT_EXPORT_MODULE()
+
++ (BOOL)requiresMainQueueSetup
+{
```
Comment (perry-mitchell) [inline ios/Crypto.m]: Move up to previous line for consistency?

File: package.json [hunk 6]
```diff
@@ -28,10 +28,9 @@
     "@tradle/react-native-http": "^2.0.0",
     "@yfuks/react-native-action-sheet": "0.0.4",
     "assert": "^1.4.1",
-    "base-64": "^0.1.0",
     "browserify-zlib": "^0.1.4",
     "buffer": "^3.6.0",
-    "buttercup": "2.0.3",
+    "buttercup": "2.0.4",
```
Comment (perry-mitchell) [inline package.json]: This should probably be `^2.0.4`

**Comment by perry-mitchell** [comment]
This can now also be rebased.. as the other testing MRs were merged 😄

**Comment by sallar** [comment]
@perry-mitchell Rebasing from master fails:

```
error: Your local changes to the following files would be overwritten by merge:
	ios/BCHelpers.m
Please commit your changes or stash them before you merge.
Aborting
error: Failed to merge in the changes.
Patch failed at 0029 Error handling for iOS
The copy of the patch that failed is found in: .git/rebase-apply/patch

Resolve all conflicts manually, mark them as resolved with
"git add/rm <conflicted_files>", then run "git rebase --continue".
You can instead skip this commit: run "git rebase --skip".
To abort and get back to the state before "git rebase", run "git rebase --abort".

fatal: ref HEAD is not a symbolic ref
```

even though master didn't change that file after my last rebase, and also there is no diff or error on the file so im not sure what to fix... weird

**Comment by perry-mitchell** [comment]
Well I guess you could compare the file with the latest in your branch, and see.. could be file props that changed too, as Mac loves changing those.

**Comment by sallar** [comment]
@perry-mitchell Updated iTerm and restarted it and it fixed it...

**Pull request merged**